### PR TITLE
API/CoW: Return copies for head and tail

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

this is something we talked about to avoid messing up references in Jupyter environments (xref https://github.com/pandas-dev/pandas/issues/48998)

### CODE REVIEW

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -25,6 +25,7 @@ Copy-on-Write improvements
 - The :class:`DataFrame` constructor, when constructing a DataFrame from a dictionary
   of Index objects and specifying ``copy=False``, will now use a lazy copy
   of those Index objects for the columns of the DataFrame (:issue:`52947`)
+- :meth:`DataFrame.head` and :meth:`DataFrame.tail` will now return copies (:issue:`54011`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v2.1.0.rst:28]: Should we clarify this as "deep copy"? With CoW, the methods already do return an object that behaves as a copy / is a shallow copy.

File: pandas/tests/copy_view/test_methods.py [hunk 2]
```diff
@@ -895,16 +895,17 @@ def test_head_tail(method, using_copy_on_write):
     df2._mgr._verify_integrity()
 
     if using_copy_on_write:
-        assert np.shares_memory(get_array(df2, "a"), get_array(df, "a"))
-        assert np.shares_memory(get_array(df2, "b"), get_array(df, "b"))
+        assert not np.shares_memory(get_array(df2, "a"), get_array(df, "a"))
```
Comment (jorisvandenbossche) [inline pandas/tests/copy_view/test_methods.py:898]: Can you add a comment here to note we explicitly deviate in head/tail methods to do an eager copy (as it now contrasts with most other tests in this file)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks!

Would there be a usecase for adding a keyword to let you not make the eager copy? That would probably only be useful if you would use head/tail for non-exploration cases, in which case you can use iloc instead (although people might prefer a method in case of method chaining)

File: doc/source/whatsnew/v2.1.0.rst (hunk 1 above)
Comment (phofl) [reply doc/source/whatsnew/v2.1.0.rst:28]: Makes sense, thx

File: pandas/tests/copy_view/test_methods.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/copy_view/test_methods.py:900]: Makes sense

**Comment by phofl** [comment]
I would not add a keyword, I am not convinced that there is a proper use-case that would require such a keyword

**Review by jorisvandenbossche — APPROVED** [review]

**Pull request merged**